### DISCUSSION TITLE
Plugin support

### DISCUSSION THREAD

**Discussion by sayanarijit** [sayanarijit/xplr#144] [Ideas]
Now I think we have enough [hacks](https://github.com/sayanarijit/xplr/wiki/Hacks) and [themes](https://github.com/sayanarijit/xplr/wiki/Themes) demo to guess what a plugin should be doing, at the simplest possible level. All these hacks and themes do one thing - overwrite the configuration. So, at the simplest level, a plugin should be able to do the same.

My basic expectations from a plugin are -

- Can overwrite the configuration safely and with the least possible overhead.
- The plugin interpreter should also be fast and easy to port.
- Some language that's easy, fun to write and nice to have proven reputation as a configuration language.

Although I have more expectations from a plugin, but for now, I think it's a good start.

With all these in mind, I think we already have an answer, i.e. [go the neovim way](https://github.com/neovim/neovim) and love [lua](http://www.lua.org/).

With [rlua](https://github.com/amethyst/rlua) and [rlua_serde](https://github.com/zrkn/rlua_serde), I already see how this can work.

So, that's my thought. I'm open to hear and discuss what the users think.

------------

Note: I also liked [zellij](https://github.com/zellij-org/zellij)'s idea of using WASM as the plugin language. But unfortunately, it doesn't fit many of my expectations listed above.

**Comment by sayanarijit** [comment]
I think, it will also be useful o attach lua functions as key handlers. With something like `LuaExec` and `LuaExecSilently`. With this, lua can read the app state directly without having to read pipe files, and can return a MsgIn without having to write to a pipe file. This will be much faster and safer than something like `BashExec` and `BashExecSilently`.

**Reply by max-baz** [reply]
What will plugin help me achieve which is not possible to implement with hacks? One clear preference for me is that with hacks I can write my scripts in any language, while with plugins it seems like I will have to go with lua and nothing else (I'm not against lua, just trying to clarify for myself, at which point would I stop trying to implement an extension using a hack and start looking into lua).

**Reply by sayanarijit** [reply]
# Why

### Portability & distribution

At the basic level, we can think of plugins as "cross-platform and more flexible hacks". With the current implementation of hacks, we are limited to what interpreters, vms or shells are installed in our system.

For e.g. we can't write a hack that executes a Python script and claim that it will work the same on any machine. Some system doesn't even come with `bash` installed. Or even if installed, we still need to worry about the version.

Lua on the other hand, can be shipped with `xplr` natively, because it's really tiny. That means, we can write a lua script and claim that it should work fine on any system. Even if there's no lua installed.

This lets us distribute our customizations for other users to use via a plugin system.

### Safety

Writing a safe bash script is really difficult. We need to escape or quote file paths properly, deal with unset and empty variables, handle errors and returncodes properly. A simple whitespace or special character in a filename can do devastating damage to our important files.

### Scriptability

With the current implementation of hacks, we are also limited to what's being exposed via the output pipes. Exposing the whole app state via pipes will be really inefficient. Although, it's possible to dump the app state as yaml, we still need `yq` to parse it which is again a portability issue. With lua, we get the whole app state passed as input. We also get inputs passed by the user when executing the plugin (think `lua require'foo'.do_something(input=bar)` in neovim). Also, we can avoid exporting global variables `e.g. $OPENER` and define it in the plugin config (think `lua require'colorizer'.setup(config=foo)`).

### Speed

Speed will be a bonus since we won't need to read files or write to files. The input and returned outputs will be converted directly by the binding logic. Also, we can surely expect lua scripts to be faster than bash scripts.

### Community

Ever since `neovim` started integrating lua natively, there's been an explosion of plugins from the community. I think we can onboard some part of this community with native lua support. If not, we can just steal from their work :smiling_imp: .

# When

As a user, you are not required to write lua. You will mostly be using basic hacks. But if you see that there's a convenient plugin that lets you replace 50 lines of hacky scripts from your config with 2-5 lines of `import`, `require` etc. you might prefer to use it instead.

**Reply by sayanarijit** [reply]
For more on why lua, you may like [this explanation](https://github.com/neovim/neovim/wiki/FAQ#why-embed-lua-instead-of-x).

**Comment by sayanarijit** [comment]
A lua alternative https://github.com/gluon-lang/gluon
